readthedocs/proxito/middleware.py

```diff
@@ -38,76 +91,41 @@ def map_host_to_project_slug(request):  # pylint: disable=too-many-return-statem
-        # Serve from the PUBLIC_DOMAIN, ensuring it looks like `foo.PUBLIC_DOMAIN`
-        if public_domain_parts == host_parts[1:]:
-            project_slug = host_parts[0]
-            request.subdomain = True
-            log.debug('Proxito Public Domain.', host=host)
-            if Domain.objects.filter(project__slug=project_slug).filter(
-                canonical=True,
-                https=True,
-            ).exists():
-                log.debug('Proxito Public Domain -> Canonical Domain Redirect.', host=host)
-                request.canonicalize = constants.REDIRECT_CANONICAL_CNAME
-            elif (
-                ProjectRelationship.objects.
-                filter(child__slug=project_slug).exists()
-            ):
-                log.debug('Proxito Public Domain -> Subproject Main Domain Redirect.', host=host)
-                request.canonicalize = constants.REDIRECT_SUBPROJECT_MAIN_DOMAIN
-            return project_slug
-
-        # TODO: This can catch some possibly valid domains (docs.readthedocs.io.com) for example
-        # But these feel like they might be phishing, etc. so let's block them for now.
-        log.warning('Weird variation on our hostname.', host=host)
-        return render(
-            request, 'core/dns-404.html', context={'host': host}, status=400
-        )
-
-    if external_domain in host:
-        # Serve custom versions on external-host-domain
-        if external_domain_parts == host_parts[1:]:
-            try:
-                project_slug, version_slug = host_parts[0].rsplit('--', 1)
-                request.external_domain = True
-                request.host_version_slug = version_slug
-                log.debug('Proxito External Version Domain.', host=host)
-                return project_slug
-            except ValueError:
-                log.warning('Weird variation on our hostname.', host=host)
-                return render(
-                    request, 'core/dns-404.html', context={'host': host}, status=400
-                )
+    project_slug, domain_object, external = _unresolve_domain(host)
+    if not project_slug:
+        # Block domains that look like ours, may be phishing.
+        if external_domain in host or public_domain in host:
+            log.warning("Weird variation on our hostname.", host=host)
+            return render(
+                request,
+                "core/dns-404.html",
+                context={"host": host},
+                status=400,
+            )
+        # Some person is CNAMEing to us without configuring a domain - 404.
+        log.debug("CNAME 404.", host=host)
+        return render(request, "core/dns-404.html", context={"host": host}, status=404)
 
-    # Serve CNAMEs
-    domain = Domain.objects.filter(domain=host).first()
-    if domain:
-        project_slug = domain.project.slug
+    # Custom domain.
+    if domain_object:
         request.cname = True
-        request.domain = domain
+        request.domain = domain_object
```

Comment:
I don't think we need to do what I'm going to propose in this PR, but, can we avoid passing things by injecting things in the `request`? In case this is not possible, or it's our best option, can we define a specific well-documented object?

```python
request.readthedocs = ReadTheDocsRequest(
  cname=bool,
  domain=Domain,
  external_domain=bool,
  host_version_slug=str,
  canonicalize=bool,
  subdomain=bool,
)
```

I think we have this kind of injections all over our code and I have no idea what are all of those possible values and how they are expected to be used. Each time I work on this code I need to grep some code to understand how they are used.

Reply:
Yeah, my goal with the new implementation is also reducing the things we inject into the request, and pass them explicitly where needed if possible, but if we still require them using an object would be another good option.

Reply:
Yea, it would be nice to standardize the data we're setting on the request. There's a few random places we're reading that data, so we need to be careful that we don't randomly break something downstream though.